cgen error when using fn callback fields

**V doctor:**

```
V full version: V 0.4.8 c941c3d
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 8 cpus, 64bit, little endian, 

getwd: D:\Documents\_Projects\_VLang\projects\35_windows_c_error
vexe: D:\Documents\_Projects\_VLang\v\v.exe
vexe mtime: 2024-12-14 02:18:22

vroot: OK, value: D:\Documents\_Projects\_VLang\v
VMODULES: OK, value: C:\Users\Mihoshi\.vmodules
VTMP: OK, value: C:\Users\Mihoshi\AppData\Local\Temp\v_0

Git version: git version 2.9.0.windows.1
Git vroot status: weekly.2024.07-1799-gc941c3d4 (1 commit(s) behind V master)
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


emcc version: Error: 'emcc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg test.v`

```v
import gg

pub type WindowResizeFn = fn (window &Window, w int, h int)

@[heap]
pub struct Window {
	id                string            = '_window_'
pub mut:
	resize_fn         WindowResizeFn    = unsafe { nil }
}

@[params]
pub struct WindowParams {
pub:
	on_resize         WindowResizeFn    = unsafe { nil }
}

fn window_resized(event gg.Event, mut w &Window) {
   window_width, window_height := 200, 100

	if w.resize_fn != WindowResizeFn(0) {
		println("fn present ${window_width} ${window_height}")
		w.resize_fn(w, window_width, window_height) // <--- C error
	}
}

fn on_event(e &gg.Event, mut w Window) {
	window_resized(e, mut w)
}

fn on_resize(window &Window, w int, h int) {
	println(" ${w} ${h}")
}

fn main() {
	e:= gg.Event{}
	mut w := &Window{
		resize_fn : on_resize
	}

	on_event(e, mut w)
}
```

**What did you expect to see?**

File should compile

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: C:/Users/Mihoshi/AppData/Local/Temp/v_0/test.01JF24XNQXBY0QCGY0NXFDSPNV.tmp.c:6463: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/Mihoshi/AppData/Local/Temp/v_0/test.01JF24XNQXBY0QCGY0NXFDSPNV.tmp.c:9898: error: expected struct or union but not 'struct main__Window *'
... (the original output was 9 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzVkNDQzNzFlMDYzMmRkMDhlMzkxYzYiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.NLekdCzoIsbwspS6N5ufXtUiLuRVl9GI-KihlyZ37yw">Huly&reg;: <b>V_0.6-21594</b></a></sub>

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -269,7 +269,7 @@ fn (mut c Checker) check_expected_call_arg(got_ ast.Type, expected_ ast.Type, la
 		// `fn foo(mut p &Expr); mut expr := Expr{}; foo(mut expr)`
 		if arg.is_mut && expected.nr_muls() > 1 && got.nr_muls() < expected.nr_muls() {
 			got_typ_str, expected_typ_str := c.get_string_names_of(got_, expected_)
-			return error('cannot use `${got_typ_str}` as `${expected_typ_str}` ${got.nr_muls()}:${expected.nr_muls()}')
+			return error('cannot use `${got_typ_str}` as `${expected_typ_str}`')
 		}
 
 		exp_sym_idx := expected.idx()

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -267,7 +267,7 @@ fn (mut c Checker) check_expected_call_arg(got_ ast.Type, expected_ ast.Type, la
 		}
 
 		// `fn foo(mut p &Expr); mut expr := Expr{}; foo(mut expr)`
-		if arg.is_mut && expected.nr_muls() > 1 && !got.is_ptr() {
+		if arg.is_mut && expected.nr_muls() > 1 && got.nr_muls() < expected.nr_muls() {
 			got_typ_str, expected_typ_str := c.get_string_names_of(got_, expected_)
 			return error('cannot use `${got_typ_str}` as `${expected_typ_str}`')
 		}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -269,7 +269,7 @@ fn (mut c Checker) check_expected_call_arg(got_ ast.Type, expected_ ast.Type, la
 		// `fn foo(mut p &Expr); mut expr := Expr{}; foo(mut expr)`
 		if arg.is_mut && expected.nr_muls() > 1 && got.nr_muls() < expected.nr_muls() {
 			got_typ_str, expected_typ_str := c.get_string_names_of(got_, expected_)
-			return error('cannot use `${got_typ_str}` as `${expected_typ_str}`')
+			return error('cannot use `${got_typ_str}` as `${expected_typ_str}` ${got.nr_muls()}:${expected.nr_muls()}')
 		}
 
 		exp_sym_idx := expected.idx()

--- a/vlib/v/checker/tests/mut_arg_different_muls_err.out
+++ b/vlib/v/checker/tests/mut_arg_different_muls_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/mut_arg_different_muls_err.vv:26:21: error: cannot use `&Window` as `&&Window` in argument 1 to `window_resized`
+   24 | 
+   25 | fn on_event(mut w Window) {
+   26 |     window_resized(mut w)
+      |                        ^
+   27 | }
+   28 |

--- a/vlib/v/checker/tests/mut_arg_different_muls_err.vv
+++ b/vlib/v/checker/tests/mut_arg_different_muls_err.vv
@@ -1,0 +1,38 @@
+pub type WindowResizeFn = fn (window &Window, w int, h int)
+
+@[heap]
+pub struct Window {
+	id string = '_window_'
+pub mut:
+	resize_fn WindowResizeFn = unsafe { nil }
+}
+
+@[params]
+pub struct WindowParams {
+pub:
+	on_resize WindowResizeFn = unsafe { nil }
+}
+
+fn window_resized(mut w &Window) {
+	window_width, window_height := 200, 100
+
+	if w.resize_fn != WindowResizeFn(0) {
+		println('fn present ${window_width} ${window_height}')
+		w.resize_fn(w, window_width, window_height)
+	}
+}
+
+fn on_event(mut w Window) {
+	window_resized(mut w)
+}
+
+fn on_resize(window &Window, w int, h int) {
+	println(' ${w} ${h}')
+}
+
+fn main() {
+	mut w := &Window{
+		resize_fn: on_resize
+	}
+	on_event(mut w)
+}

--- a/vlib/v/tests/conditions/ifs/in_expression_test.v
+++ b/vlib/v/tests/conditions/ifs/in_expression_test.v
@@ -338,5 +338,5 @@ fn in_both_mut_ref(mut arr []&Bar, mut bar &Bar) {
 fn test_in_both_mut_and_ref() {
 	mut arr := []&Bar{}
 	mut bar := &Bar{}
-	in_both_mut_ref(mut &arr, mut bar)
+	in_both_mut_ref(mut &arr, mut &bar)
 }

--- a/vlib/v/tests/fns/fn_call_mut_ref_args_test.v
+++ b/vlib/v/tests/fns/fn_call_mut_ref_args_test.v
@@ -17,7 +17,7 @@ fn init_vm2(mut head &Client) {
 
 fn test_fn_call_mut_ref_args() {
 	mut head := &Client{}
-	init_vm1(mut head)
-	init_vm2(mut head)
+	init_vm1(mut &head)
+	init_vm2(mut &head)
 	assert true
 }

--- a/vlib/v/tests/fns/fn_mut_arg_of_sumtype_ref_test.v
+++ b/vlib/v/tests/fns/fn_mut_arg_of_sumtype_ref_test.v
@@ -9,7 +9,7 @@ pub type Expr = Expr1 | Expr2
 pub struct Gen {
 }
 
-fn (mut g Gen) foo(mut expr &Expr1) string {
+fn (mut g Gen) foo(mut expr Expr1) string {
 	return '${expr}'
 }
 

--- a/vlib/v/tests/options/option_reference_params_test.v
+++ b/vlib/v/tests/options/option_reference_params_test.v
@@ -23,6 +23,6 @@ fn bang(mut logr ?&Logger, this string) {
 
 fn test_option_reference_params() {
 	mut logr := new_logger()
-	bang(mut logr, 'bang!')
+	bang(mut &logr, 'bang!')
 	assert true
 }

